streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -368,203 +299,166 @@ public UUID processId() {
-        for (final StandbyTask standbyTask : running) {
-            checkpointedOffsets.putAll(standbyTask.checkpointedOffsets());
-        }
-
-        log.debug("Assigning and seeking restoreConsumer to {}", checkpointedOffsets);
-        restoreConsumerAssignedStandbys = true;
-        restoreConsumer.assign(checkpointedOffsets.keySet());
-        for (final Map.Entry<TopicPartition, Long> entry : checkpointedOffsets.entrySet()) {
-            final TopicPartition partition = entry.getKey();
-            final long offset = entry.getValue();
-            if (offset >= 0) {
-                restoreConsumer.seek(partition, offset);
-            } else {
-                restoreConsumer.seekToBeginning(singleton(partition));
+        final List<Task> restoringTasks = new LinkedList<>();
+        for (final Task task : tasks.values()) {
+            task.initializeIfNeeded();
+            if (task.state() == RESTORING) {
+                restoringTasks.add(task);
             }
         }
-    }
 
-    public void setRebalanceInProgress(final boolean rebalanceInProgress) {
-        this.rebalanceInProgress = rebalanceInProgress;
-    }
-
-    public void setClusterMetadata(final Cluster cluster) {
-        this.cluster = cluster;
-    }
-
-    public void setPartitionsByHostState(final Map<HostInfo, Set<TopicPartition>> partitionsByHostState) {
-        this.streamsMetadataState.onChange(partitionsByHostState, cluster);
-    }
-
-    public void setPartitionsToTaskId(final Map<TopicPartition, TaskId> partitionsToTaskId) {
-        this.partitionsToTaskId = partitionsToTaskId;
-    }
-
-    public void setAssignmentMetadata(final Map<TaskId, Set<TopicPartition>> activeTasks,
-                                      final Map<TaskId, Set<TopicPartition>> standbyTasks) {
-        addedActiveTasks.clear();
-        for (final Map.Entry<TaskId, Set<TopicPartition>> entry : activeTasks.entrySet()) {
-            if (!assignedActiveTasks.containsKey(entry.getKey())) {
-                addedActiveTasks.put(entry.getKey(), entry.getValue());
+        if (!restoringTasks.isEmpty()) {
+            changelogReader.restore();
+            final Set<TopicPartition> restored = changelogReader.completedChangelogs();
+            for (final Task task : restoringTasks) {
+                if (restored.containsAll(task.changelogPartitions())) {
+                    task.startRunning();
+                }
             }
         }
 
-        addedStandbyTasks.clear();
-        for (final Map.Entry<TaskId, Set<TopicPartition>> entry : standbyTasks.entrySet()) {
-            if (!assignedStandbyTasks.containsKey(entry.getKey())) {
-                addedStandbyTasks.put(entry.getKey(), entry.getValue());
+        boolean allRunning = true;
+        for (final Task task : tasks.values()) {
+            // TODO, can we make StandbyTasks partitions always empty (since they don't process any inputs)?
```

Comment:
Use TODO K9913?

Reply:
I intend to resolve these before merging this PR.